### ISSUE TITLE
Wallet Backend builds and tests on Windows

### ISSUE THREAD

**Issue by KtorZ** [cardano-foundation/cardano-wallet#703]
# Context

`cardano-wallet` ultimately results in two binaries as detailed in the root [.cabal](https://github.com/input-output-hk/cardano-wallet/blob/master/cardano-wallet.cabal):

- cardano-wallet-http-bridge
- cardano-wallet-jormungandr

Both are command-lines and web server doing "wallet backend stuff", but working with different chain producers. 

# Decision

At the moment, our CI only produces binary artifacts for linux machines yet, most are using Windows and OSX machines so we ought to be able to produce executables for these targets too. Ideally, our test suites should also run on both platform to ensure that code produced works as expected on these platforms.

The focus is given on `cardano-wallet-jormungandr` for `cardano-wallet-http-bridge` will be removed soon enough and will not see the lights of production.

# Acceptance Criteria

- [x] `cardano-wallet-jormungandr` _must_ compile to Windows
- [x] `cardano-wallet-core:test:unit` _must_ pass on Windows
- [x] `cardano-wallet-cli:test:unit` _must_ pass on Windows
- [x] `cardano-wallet-launcher:test:unit` _must_ pass on Windows
- [x] `bech32:test:unit` _must_ pass on Windows
- [x] `cardano-wallet-jormungandr:test:unit` _must_ pass on Windows
- [ ] `cardano-wallet-jormungandr:test:integration` _must_ pass on Windows
- [x] `cardano-wallet-core:benchmark:db` _should_ pass on Windows

---

# Development Plan

1. Get all tests working under nix-build for Linux.
2. Get cross-compiled tests passing under Wine.
3. Manually test .exe on a Windows 10 VM.

- [x] Run tests in release build ⇒ #766.
- [x] Update Haskell.nix version to unbreak windows cross build ⇒ #888.
- [x] Include windows build of Jormungandr into releases ⇒ #906.
- [x] Fix failures in sqlite tests on Windows ⇒ #918.
- [x] ~Fix failure in launch command on Windows.~ (is no longer an issue)
- [x] Fix the UTF-8 text issues on Windows ⇒ #914.
- [x] Add code which tells the windows command prompt to interpret ANSI color codes ⇒ #957.
- [x] Ensure test data is present in windows testing bundle ⇒ #958.
- [x] Test suite `cardano-wallet-core:test:unit` ([failing tests](https://gist.github.com/rvl/d27fb803a670e91a6826d104e8a8e8d5))
    - [x] golden tests fail because of filenames ⇒ #981 
    - [x] Some SQLite tests fail, presumably because of windows file locking ⇒ #971.
    - [x] Error path tests in `Cardano.Wallet.Api.ServerSpec` ⇒ #971
        - Network.Socket.gai_strerror not supported: 11001
        - Cannot assign requested address (WSAEADDRNOTAVAIL)
        - Assertion in privileged port test does not apply to windows
        - "Port in use" test case fails to fail. This is a windows "feature" ⇒ #1016
    - [x] Middleware.LoggingSpec tests fail because number of log messages are not of the expected number ⇒ #974.
    - [x] Windows prompts about firewall things. The tests need to be allowed to make connections.
- [x] `cardano-wallet-launcher:test:unit` fails due to use of posix commands in tests ⇒ #959.
- [x] Database benchmark fails with "Sharing violation" ⇒ #975.
- [x] Killing the wallet launcher doesn't kill the underlying Jörmungandr process on Windows? ⇒ #986.
- [ ] Test suite `cardano-wallet-jormungandr:test:integration` can run under Wine ⇒ #1011.
- [ ] Identify all other test failures and fix.

# QA 

### Checking for windows builds and tests in CI

1. Builds for the master branch are shown in the [cardano-wallet jobset](https://hydra.iohk.io/jobset/Cardano/cardano-wallet#tabs-jobs).
2. Ensure that `x86_64-pc-mingw32` jobs for the packages listed in the AC are green.
3. Check that the [`required`](https://hydra.iohk.io/job/Cardano/cardano-wallet/required) job is green.
4. Check the _Constituents_ of the `required` job. These should include `x86_64-pc-mingw32` jobs for the packages listed in the AC.

### Running tests on windows

1. Download the [Windows testing bundle](https://hydra.iohk.io/job/Cardano/cardano-wallet/cardano-wallet-jormungandr-win64/latest) onto a Windows system.
2. Extract to zip archive to somewhere (e.g. Desktop).
3. Ensure old data in `c:\cardano-wallet-jormungandr` is removed.
4. Use the example batch files `launch.bat`, `cw.bat`, `jm.bat`.
5. Or you can invoke `cardano-wallet-jormungandr.exe` yourself.
6. Run all tests with `tests.bat`.
7. Or you can run test suites individually with their `.exe` files.

### Building windows testing bundle locally

- See [Building Wiki](https://github.com/input-output-hk/cardano-wallet/wiki/Building#nix-build) for background info.
- `nix-build release.nix -A cardano-wallet-jormungandr-tests-win64 && ls ./result`

### Running tests with Wine

```
nix-env -f '<nixpkgs>' -iA pkgs.wineWowPackages.minimal  # Install Wine if you don't already have it.
nix-build release.nix -A x86_64-pc-mingw32.tests.cardano-wallet-core.unit.x86_64-linux  # Build a test suite
wine ./result/cardano-wallet-core-2019.11.7/unit.exe --match ...
```

**Comment by rvl** [comment]
## How to test cardano-wallet-jormungandr on Windows

1. If you don't already have it, set up Windows 10 in VirtualBox.
   You can use the Powershell script and instructions at
   [iohk-ops/windows/iohk-setup.ps1](https://github.com/input-output-hk/iohk-ops/blob/master/windows/iohk-setup.ps1).
   Just delete everything in the script starting from where it installs GHC.

   Also make sure you have the
   [Microsoft Visual C++ Redistributable for Visual Studio 2015](https://chocolatey.org/packages/vcredist-all)
   (or maybe it was 2017, I forget now).

2. Set up a shared folder between Linux and Windows.
   Let's call it `~/iohk/win_shared` or `E:`.

3. Under the `cardano-wallet` repo, run:

   ```
   nix-build release.nix -A cardano-wallet-jormungandr-win64
   cp -vf result/*.zip ~/iohk/win_shared
   ```

   Alternatively, download the prebuilt distribution from Hydra:
   [cardano-wallet-jormungandr-win64](https://hydra.iohk.io/job/Cardano/cardano-wallet/cardano-wallet-jormungandr-win64/latest) | [zip download](https://hydra.iohk.io/job/Cardano/cardano-wallet/cardano-wallet-jormungandr-win64/latest/download-by-type/file/binary-dist)

4. On Windows, open `E:` in explorer, then right-click on the zip file
   and select "Extract All...". Extract it to somewhere like
   `C:\cardano-wallet-jormungandr`.

5. Open up two command prompt windows (Start→Run `cmd.exe`) and `cd`
   to where you extracted the files.

6. In the first window, type `jm` [ENTER].

7. In the second window, type `cw` [ENTER].

**Comment by KtorZ** [comment]
@rvl @disasm

Can someone give a clear albeit concise status update on Windows builds?

In particular, answers to the following questions:

1. Can we build a Windows executable that runs, without crashing, the wallet server?
2. Can we run our unit tests on wine or Windows directly?
3. Can we run our integration scenarios on wine or Windows?
4. If it exists some build artifacts that are produced for Windows, where can they be found?
5. If any answer to the questions above is "no", what are the current blockers or items being worked on?

I've got the impression that many things are happening in background but I can't clearly see what because nothing seem to be really tracked and discussions are happening.. Somewhere else. There are PR pending with no descriptions, and it's very hard to follow.

Thank you very much :pray:

**Comment by KtorZ** [comment]
@rvl How are we doing with wine tests execution here? Wine doesn't work for me locally for a.. mysterious reason and as far as I can see, the Hydra job doesn't really work either.

**Comment by rvl** [comment]
We are waiting for execution of wine tests to be put back - input-output-hk/haskell.nix#316. Currently it builds the test suite for windows but does not run it, so it's always green in Hydra.

I don't know what version and configuration of wine you were using. But the version used in the nix build is capable of running the test suite and `jormungandr.exe`.

**Comment by KtorZ** [comment]
I am using the "default" one from aptitude on Ubuntu 18.04, so it's probably not the most recent version as usual ^^" ... 
I'll try getting it installed through nix and see.

**Comment by piotr-iohk** [comment]
> @KtorZ KtorZ added this to QA in cardano-wallet 2 hours ago

Does it mean that tests are running on wine in CI?

**Comment by KtorZ** [comment]
Whoops. My bad, I messed up when setting up the GH board

**Comment by KtorZ** [comment]
closing this one. Tests are now running in hydra:

### unit

- https://hydra.iohk.io/job/Cardano/cardano-wallet/x86_64-pc-mingw32.checks.text-class.unit.x86_64-linux
- https://hydra.iohk.io/job/Cardano/cardano-wallet/x86_64-pc-mingw32.checks.cardano-wallet-launcher.unit.x86_64-linux
- https://hydra.iohk.io/job/Cardano/cardano-wallet/x86_64-pc-mingw32.checks.cardano-wallet-jormungandr.unit.x86_64-linux
- https://hydra.iohk.io/job/Cardano/cardano-wallet/x86_64-pc-mingw32.checks.cardano-wallet-core.unit.x86_64-linux
- https://hydra.iohk.io/job/Cardano/cardano-wallet/x86_64-pc-mingw32.checks.cardano-wallet-cli.unit.x86_64-linux
- https://hydra.iohk.io/job/Cardano/cardano-wallet/x86_64-pc-mingw32.checks.bech32.bech32-test.x86_64-linux

### integration

- https://hydra.iohk.io/job/Cardano/cardano-wallet/x86_64-pc-mingw32.checks.cardano-wallet-jormungandr.integration.x86_64-linux



Some jobs are still failing but this is tracked by #1115
